lib/src/components/joystick/joystick_action.dart

```diff
@@ -11,65 +11,58 @@ import '../../../game.dart';
 import '../../extensions/offset.dart';
 import '../../extensions/rect.dart';
 import '../../extensions/vector2.dart';
-import '../../sprite.dart';
 import 'joystick_component.dart';
+import 'joystick_element.dart';
 import 'joystick_events.dart';
-import 'joystick_utils.dart';
 
 enum JoystickActionAlign { topLeft, bottomLeft, topRight, bottomRight }
 
 class JoystickAction extends BaseComponent with Draggable, HasGameRef {
   final int actionId;
-  final Sprite sprite;
-  final Sprite spritePressed;
-  final Sprite spriteBackgroundDirection;
   final double size;
-  final double sizeFactorBackgroundDirection;
+  final double _sizeBackgroundDirection;
   final EdgeInsets margin;
   final JoystickActionAlign align;
   final bool enableDirection;
-  final Color color;
-  final double opacityBackground;
-  final double opacityKnob;
 
   bool isPressed = false;
-  Rect _rectAction;
-  Rect _rectBackgroundDirection;
   bool _dragging = false;
-  Sprite _spriteAction;
   Vector2 _dragPosition;
-  final Paint _paintBackground;
-  final Paint _paintAction;
-  final Paint _paintActionPressed;
-  final double _sizeBackgroundDirection;
   double _tileSize;
 
+  JoystickElement backgroundDirection;
+  JoystickElement action;
+  JoystickElement actionPressed;
+
   JoystickController get joystickController => parent as JoystickController;
 
   JoystickAction({
     @required this.actionId,
-    this.sprite,
-    this.spritePressed,
-    this.spriteBackgroundDirection,
+    JoystickElement backgroundDirection,
+    JoystickElement action,
+    JoystickElement actionPressed,
     this.enableDirection = false,
     this.size = 50,
-    this.sizeFactorBackgroundDirection = 1.5,
     this.margin = EdgeInsets.zero,
-    this.color = Colors.blueGrey,
     this.align = JoystickActionAlign.bottomRight,
-    this.opacityBackground = 0.5,
-    this.opacityKnob = 0.8,
-  })  : _spriteAction = sprite,
-        _sizeBackgroundDirection = sizeFactorBackgroundDirection * size,
-        _paintBackground = Paint()
-          ..color = color.withOpacity(opacityBackground)
-          ..style = PaintingStyle.fill,
-        _paintAction = Paint()
-          ..color = color.withOpacity(opacityKnob)
-          ..style = PaintingStyle.fill,
-        _paintActionPressed = Paint()
-          ..color = color.withOpacity(opacityBackground)
-          ..style = PaintingStyle.fill {
+    double sizeFactorBackgroundDirection = 1.5,
+    Color color = Colors.blueGrey,
+    double opacityBackground = 0.5,
+    double opacityKnob = 0.8,
```

Comment:
in a followup we could even remove this as you can provide your own Element.paint

Reply:
(in fact I prefer that as it simplifies things)